src/duration.rs

```diff
@@ -589,10 +625,55 @@ mod tests {
         assert_eq!(Duration::microseconds(1001).num_milliseconds(), 1);
         assert_eq!(Duration::microseconds(-999).num_milliseconds(), 0);
         assert_eq!(Duration::microseconds(-1001).num_milliseconds(), -1);
-        assert_eq!(Duration::milliseconds(i64::MAX).num_milliseconds(), i64::MAX);
-        assert_eq!(Duration::milliseconds(-i64::MAX).num_milliseconds(), -i64::MAX);
-        assert_eq!(MAX.num_milliseconds(), i64::MAX);
-        assert_eq!(MIN.num_milliseconds(), -i64::MAX);
+    }
+    #[test]
+    fn test_duration_milliseconds_max_allowed() {
+        // The maximum number of milliseconds acceptable through the constructor is
+        // equal to the number that can be stored in a Duration.
+        let duration = Duration::milliseconds(i64::MAX);
+        assert_eq!(duration.num_milliseconds(), i64::MAX);
+        assert_eq!(
+            duration.secs as i128 * 1_000_000_000 + duration.nanos as i128,
+            i64::MAX as i128 * 1_000_000
+        );
+    }
+    #[test]
+    fn test_duration_milliseconds_max_overflow() {
+        // Here we ensure that trying to add one millisecond to the maximum storable
+        // value will fail.
+        assert!(Duration::milliseconds(i64::MAX).checked_add(&Duration::milliseconds(1)).is_none());
+    }
+    #[test]
+    fn test_duration_milliseconds_min_allowed() {
+        // The minimum number of milliseconds acceptable through the constructor is
+        // not equal to the number that can be stored in a Duration - there is a
+        // difference of one (i64::MIN vs -i64::MAX).
+        let duration = Duration::milliseconds(-i64::MAX);
+        assert_eq!(duration.num_milliseconds(), -i64::MAX);
+        assert_eq!(
+            duration.secs as i128 * 1_000_000_000 + duration.nanos as i128,
+            -i64::MAX as i128 * 1_000_000
+        );
+    }
+    #[test]
+    fn test_duration_milliseconds_min_underflow() {
```

Comment:
Can you adjust the tests in this commit so that all tests pass? Or just reorder the commit to after the fix? Our goal is to make each commit have all tests passing to make dissecting easier.

Reply:
Understood - the purpose here was to test existing code, _then_ apply the fix, but that makes more sense really when there's a concept of whole-branch arcs. As all my commits will be rebase-merged, I can understand that as part of that stream of commits you'd prefer to have each one passing, even if that means concealing an issue that would in fact fail. What I will do is move the test from the commit adding the tests to the commit with the fix, which somewhat clouds the narrative but does preserve the ideal of every key point on the `main` branch passing 🙂

Reply:
I've done that - in the end I chose to do it by marking the failing test as to be ignored, and then un-ignoring it when the fix was applied. I trust that strikes the right balance? It means all tests pass in each commit, but still preserves the narrative.